Add support for configuring JSONDecoder.keyDecodingStrategy in Xcode 9.3

mpvosseller opened the pull request:
Starting with Xcode 9.3 (currently in beta) there is a nice new property in `JSONDecoder` called `keyDecodingStrategy` that I would like to customize (once it is out of beta). https://benscheirman.com/2018/02/swift-4-1-keydecodingstrategy

This pull request is intended to explore and discuss the options and tradeoffs of ways to support this.

**Option 1** - Add a new optional `keyDecodingStrategy` argument to `load`. The main problem with this is that it will break when building with versions prior to Xcode 9.3. (How long should we support Xcode 9.2?). It also feels a _little_ dirty to accept multiple arguments to configure a single `JSONDecoder` but maybe that is OK.

**Option 2** - Add a new optional `JSONDecoder` parameter to the existing `load` function. This would be a little weird because it would allow passing both a `dateDecodingStrategy` and a `JSONDecoder` and it wouldn't be super obvious what we would do if the user provided both.

**Option 3** - The attached code just creates a second `load` method that accepts an optional `JSONDecoder` argument instead of `dateDecodingStrategy`. This isn't practical though because the compiler will complain if you try to call `load` without the second argument (it would be an ambiguous function call)

**Option 4** - Ideally we would do Option 3 but also delete the old `load` method. The problem with this is that it breaks backwards source compatibility and I wasn't sure how strict this project is with that. Existing API users who previously provided a `dateDecodingStrategy` argument would need to update their code to instead pass a `JSONDecoder`.

**Option 5** - The simple solution is Option 3 but rename the new `load` method to something else . Coming up with a new name is the hard part. :-)

Any other thoughts / ideas? WDYT?

mpvosseller committed:
Accept JSONDecoder as an optional argument in load.

pietbrauer commented:
I agree, providing a decoder is a cleaner option than passing in all the parameters. Can you make that work?

mpvosseller committed:
Merge branch 'master' into json-decoder-improvements

mpvosseller committed:
improves backwards source compatibility

pietbrauer commented on RequestKit/Router.swift:
```diff
@@ -47,8 +47,8 @@ public protocol Router {
     func loadJSON<T: Codable>(_ session: RequestKitURLSession, expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol?
-    func load<T: Codable>(_ session: RequestKitURLSession, dateDecodingStrategy: JSONDecoder.DateDecodingStrategy?, expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol?
-
+    func load<T: Codable>(_ session: RequestKitURLSession, dateDecodingStrategy: JSONDecoder.DateDecodingStrategy, expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol?
```
Maybe worth deprecating the old method?

mpvosseller commented:
How do you feel about this (see attached code):

1- I removed the existing `load` method
```
public func load<T: Codable>(_ session: RequestKitURLSession = URLSession.shared, dateDecodingStrategy: JSONDecoder.DateDecodingStrategy? = nil, expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol? {
```

2- I added a new `load` method that is similar but it that replaces the optional `dateDecodingStrategy` parameter with a `JSONDecoder` parameter.

```
 public func load<T: Codable>(_ session: RequestKitURLSession = URLSession.shared, decoder: JSONDecoder = JSONDecoder(), expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol?
```

3- To improve backwards source compatibility I also added a new `load` method like the original but makes the `dateDecodingStrategy` parameter non-optional.
 
```
func load<T: Codable>(_ session: RequestKitURLSession = URLSession.shared, dateDecodingStrategy: JSONDecoder.DateDecodingStrategy, expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol?
```
The downsides are that in breaks source compatibility in some cases. It breaks for those who previously were calling `load` while explicitly passing a `dateDecodingStrategy` argument that was an optional (or nil). (Shouldn't be many or any). It also breaks source compatibility for anyone adopting the `Router` protocol directly.

pietbrauer commented:
If we make a breaking change we can also remove the old method completely. Why not keep the optional `dateDecodingStrategy` and just deprecate the old `load` method?

mpvosseller committed:
make dateDecodingStrategy an optional type

mpvosseller commented:
Actually, I think I just figured out how to support this without breaking backwards compatibility. The attached code leaves the current `load` method as is but provides a new default implementation of that function which no longer provides a default value for `dateDecodingStrategy`. This fixes the ambiguity between functions with default parameters and I think it will work for everyone.

So the new implementation just is:

1) adds a new `load` method that accepts a  `JSONDecoder` (instead of a `dateDecodingStrategy`) and looks like this:
```
func load<T: Codable>(_ session: RequestKitURLSession, decoder: JSONDecoder, expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol?
```

2) provides new reasonable default implementations for the both `load` methods.

I think this will work.

WDYT?

mpvosseller commented:
@pietbrauer I saw you approved the changes but didn't merge. OK to merge?
